nautobot/dcim/filters.py

```diff
@@ -299,22 +295,15 @@ class Meta:
             "outer_unit",
         ]
 
-    def search(self, queryset, name, value):
-        if not value.strip():
-            return queryset
-        return queryset.filter(
-            Q(name__icontains=value)
-            | Q(facility_id__icontains=value)
-            | Q(serial__icontains=value.strip())
-            | Q(asset_tag__icontains=value.strip())
```

Comment:
How are/will we handle this?

Reply:
Just always strip it. There is no use-case for trailing spaces in search terms.

Reply:
A lot of Christmas striping. I missed that you do `value = value.strip()` earlier in the new filter. 👍

Reply:
Thinking about this a bit more - I could argue that leading/trailing whitespace should *always* be preserved in `icontains` matches, but should always be *stripped* when doing `iexact` matches - thoughts on that idea?

Reply:
> Thinking about this a bit more - I could argue that leading/trailing whitespace should _always_ be preserved in `icontains` matches, but should always be _stripped_ when doing `iexact` matches - thoughts on that idea?

If we change `serial` and `asset_tag` to be `exact` searches and it wouldn't break functionality/expectation for users, then yes I am fine with that.

Otherwise I think we should implement some way that when defining the predicates you can optionally provide a value preprocessor to run through.

Something like:

```py
filter_predicates={
    "name": "icontains",
    "comments": "icontains",
    "serial": {
        "lookup_expression": "icontains",
        "preprocessor": lambda v : v.strip
    }
},

```

Reply:
To split the difference, stripping on `icontains` and not on everything else keeps it simply without having to account for variadic processing of `filter_predicates` field values.

I'm good with establishing assertions on search patterns. My assertion that "there is no use-case for trailing spaces in search terms" was based on the fact that the test code is not exercising any such cases. Another nice benefit of centralizing this functionality is being more exhaustive in testing the actual use-cases.

Reply:
> To split the difference, stripping on `icontains` and not on everything else keeps it simply without having to account for variadic processing of `filter_predicates` field values.

I'm thinking the inverse of this. I can see places where keeping whitespace on `icontains` searches would be desirable (see: https://github.com/nautobot/nautobot/pull/1752#discussion_r865841565) but on `exact` we would ideally want to strip (@glennmatthews had pointed out copy🍝 of serial numbers).

My example probably wasn't the best, it should have been:
```py
filter_predicates={
    "name": "icontains",
    "comments": "icontains",
    "serial": {
        "lookup_expression": "iexact",
        "preprocessor": lambda v : v.strip
    }
},
```

Reply:
In this case, there's the value preprocessor, and then there's the "type" I was working with. This appears to replace the need for them both. For example consider `Site.asn`: 

```python
filter_predicates={
    "asn": {
        "lookup_expression": "iexact",
        "preprocessor": int
    }
},
```

This pattern appears to do the same as `{asn:int": "exact"}` but with `preprocessor` being any callable, it also opens the door to a `lambda` or whatever else?

Reply:
> In this case, there's the value preprocessor, and then there's the "type" I was working with. This appears to replace the need for them both. For example consider `Site.asn`:
> 
> ```python
> filter_predicates={
>     "asn": {
>         "lookup_expression": "iexact",
>         "preprocessor": int
>     }
> },
> ```
> 
> This pattern appears to do the same as `{"asn:int": "iexact"}` but with `preprocessor` being any callable, it also opens the door to a `lambda` or whatever else (e.g. `"preprocessor": lambda v: int(v.strip())`)
> 

I like this. IMO as long as `preprocessor` is a callable with a single input we should be GTG.

> Also, I think it would be more wise to just have the preprocessor be called itself, and whatever goes on in the lambda should already be called.

What are you suggesting here? That the code in `SearchFilter` do something like:

```py
#...
for filter_key, filter_options in filter_predicates:
    #...
    value = filter_options['preprocessor'](value)
```

Or something else?

Reply:
I went ahead and implemented exactly this pattern in the latest commit (fcf7f14a)

Reply:
I'm still voting for the default preprocessor for `[i]exact` to be `str.strip` while the default preprocessor for `[i]contains` would be a no-op.